### PR TITLE
Connecting to offchain data - Updated Oracles lesson

### PR DESCRIPTION
**Pull request by 0xCipherCoder**
### Summary of Changes

- Updated code snippets with the latest anchor and switchboard-solana version
- Fixed content with updated version changes
- Fixed grammar and styling
- Fixed as per guidelines

Fixes #
Unboxed PRs
Burry Escrow code updated - https://github.com/solana-developers/burry-escrow/pull/4

@mikemaccana I need your help in creating an additional branch for the challenge solution.

### CODE REVIEW

File: content/courses/connecting-to-offchain-data/oracles.md [hunk 1]
```diff
@@ -12,255 +12,252 @@ description: Access real-world data inside a Solana program.
...
-oracle operator acts maliciously and attempts to change the operation of the
-approved code, a data quote verification will fail. This allows Switchboard
-oracles to operate beyond quantitative value reporting, such as functions --
-running offchain custom and confidential computations.
+Solana has a rich ecosystem of oracle providers. Some notable ones include
+[Pyth Network](https://pyth.network), [Switchboard](https://switchboard.xyz),
+[Chainlink](https://chain.link), and
+[DIA](https://www.diadata.org/solana-price-oracles/). Two of the most well-known
+are [Pyth](https://pyth.network) and [Switchboard](https://switchboard.xyz),
+each with unique design choices.
```
Comment (mikemaccana) [inline content/courses/connecting-to-offchain-data/oracles.md:115]: There's some duplication here from the old text. I suggest:

```
Solana has a rich ecosystem of oracle providers. Some notable ones include:

 - [**Pyth**](https://www.pyth.network/price-feeds) ...
 - [**Switchboard**](https://example.com) ...
 - [**Chainlink**](https://example.com) ...
 - [DIA](https://www.diadata.org/solana-price-oracles/) ...
 
 We'll use Switchboard in this lesson, but the concepts are common to most oracles and you should pick the oracle that best meets your needs. 
```

File: content/courses/connecting-to-offchain-data/oracles.md [hunk 2]
```diff
@@ -12,255 +12,252 @@ description: Access real-world data inside a Solana program.
...
-
-For example, there is a Switchboard-sponsored
-[BTC_USD feed](https://app.switchboard.xyz/solana/devnet/feed/8SXvChNYFhRq4EZuZvnhjrB3jJRQCv4k3P4W6hesH3Ee).
-This feed is available on Solana devnet/mainnet with pubkey
-`8SXvChNYFhRq4EZuZvnhjrB3jJRQCv4k3P4W6hesH3Ee`. It provides the current price of
-Bitcoin in USD onchain.
-
-The actual onchain data for a Switchboard feed account looks a little like this:
+Switchboard oracles store data on Solana using data feeds, also called
+aggregators. These data feeds consist of multiple jobs that are aggregated to
```
Comment (mikemaccana) [inline content/courses/connecting-to-offchain-data/oracles.md:150]:
```suggestion
*aggregators*. These data feeds consist of multiple jobs that are aggregated to
```

File: content/courses/connecting-to-offchain-data/oracles.md [hunk 3]
```diff
@@ -303,85 +300,84 @@ Some relevant fields and configurations on the `AggregatorAccountData` type are:
...
 ...
 
 #[derive(Accounts)]
 pub struct ConsumeDataAccounts<'info> {
-	// pass in data feed account and deserialize to AggregatorAccountData
-	pub feed_aggregator: AccountLoader<'info, AggregatorAccountData>,
-	...
+  // Pass in data feed account and deserialize to AggregatorAccountData
+  pub feed_aggregator: AccountLoader<'info, AggregatorAccountData>,
+  ...
```
Comment (mikemaccana) [inline content/courses/connecting-to-offchain-data/oracles.md:354]: Formatting is off here.

File: content/courses/connecting-to-offchain-data/oracles.md [hunk 4]
```diff
@@ -303,85 +300,84 @@ Some relevant fields and configurations on the `AggregatorAccountData` type are:
...
-`AggregatorAccountData`, the account uses what's called zero copy. This in
-combination with `AccountLoader` prevents the account from being loaded into
-memory and gives our program direct access to the data instead. When using
-`AccountLoader` we can access the data stored in the account in one of three
-ways:
+Notice that we use the `AccountLoader` type here instead of the regular
+`Account` type to deserialize the aggregator account. Due to the size of
+`AggregatorAccountData`, the account uses zero-copy deserialization. This,
+combined with `AccountLoader`, prevents the account from being fully loaded into
+memory, allowing our program to access the data directly.
```
Comment (mikemaccana) [inline content/courses/connecting-to-offchain-data/oracles.md:362]: This still feels vague. "We can't load the data, so we can load it directly"

I *think* this is what it means:

>  Zero copy,combined with `AccountLoader`, prevents the account from being fully loaded into
memory. Instead our program can access individual data items within the account directly.

If that's correct, please update the text to read something like this (in your own words, whatever makes this clearest).

If that's not correct, please reply here!

File: content/courses/connecting-to-offchain-data/oracles.md [hunk 5]
```diff
@@ -414,27 +410,34 @@ You can also view the current value stored in an `AggregatorAccountData` account
...
+// Pass switchboard program object and feed pubkey into AggregatorAccount constructor
+aggregatorAccount = new AggregatorAccount(
+  switchboardProgram,
+  SOL_USD_SWITCHBOARD_FEED,
+);
 
-// fetch latest SOL price
-const solPrice: Big | null = await aggregatorAccount.fetchLatestValue()
+// Fetch latest SOL price
+const solPrice: Big | null = await aggregatorAccount.fetchLatestValue();
```
Comment (mikemaccana) [inline content/courses/connecting-to-offchain-data/oracles.md:438]: Is `Big` the same as `BN`? I can't see the import.

File: content/courses/connecting-to-offchain-data/oracles.md [hunk 6]
```diff
@@ -473,21 +476,21 @@ Once you have found a feed that fits your needs, you still need to make sure
...
+of the account passed in matches what it should be. One way to do this is to
+hard code the address in the program somewhere and use account constraints to
+verify the address passed in matches what is expected.
 
 ```rust
 use {
   anchor_lang::prelude::*,
   solana_program::{pubkey, pubkey::Pubkey},
-	switchboard_v2::{AggregatorAccountData},
+  switchboard_solana::{AggregatorAccountData},
```
Comment (mikemaccana) [inline content/courses/connecting-to-offchain-data/oracles.md:493]: Format this (just paste into deleteme.rs in vscode, save it so it formats, then paste back)

File: content/courses/connecting-to-offchain-data/oracles.md [hunk 7]
```diff
@@ -613,139 +616,144 @@ use {
...
+    └── src
+        ├── constants.rs
+        ├── errors.rs
+        ├── instructions
+        │   ├── deposit.rs
+        │   ├── mod.rs
+        │   └── withdraw.rs
+        ├── lib.rs
+        └── state.rs
+```
```
Comment (mikemaccana) [inline content/courses/connecting-to-offchain-data/oracles.md:726]: No feedback here just wanted to say I really appreciate the attention to detail, as always.

File: content/courses/connecting-to-offchain-data/oracles.md [hunk 8]
```diff
@@ -754,39 +762,45 @@ mod burry_oracle_program {
...
-We will also be defining our PDA seed of `"MICHAEL BURRY"` and our hardcoded
-SOL_USD oracle pubkey `SOL_USDC_FEED`.
+- `unlock_price`: The price of SOL in USD at which withdrawals are allowed
+  (e.g., hard-coded to $21.53).
+- `escrow_amount`: Tracks the amount of lamports held in the escrow account.
 
-```rust
-// in state.rs
+```rust filename="state.rs"
+// Inside state.rs file
```
Comment (mikemaccana) [inline content/courses/connecting-to-offchain-data/oracles.md:775]:
```suggestion
```

Comment (mikemaccana) [reply content/courses/connecting-to-offchain-data/oracles.md:775]: We already have the filename above.

File: content/courses/connecting-to-offchain-data/oracles.md [hunk 9]
```diff
@@ -1161,241 +1175,236 @@ Note that there can only be one escrow per user, so the above order matters.
...
+      throw new Error("Aggregator holds no value");
+    }
+    const unlockPrice = solPrice.plus(PRICE_OFFSET).toNumber();
+    await createAndVerifyEscrow(unlockPrice);
+  });
 
-    // derive escrow address
-    const [escrowState] = await anchor.web3.PublicKey.findProgramAddressSync(
-      [Buffer.from("MICHAEL BURRY"), payer.publicKey.toBuffer()],
+  it("fails to withdraw while price is below UnlockPrice", async () => {
```
Comment (mikemaccana) [inline content/courses/connecting-to-offchain-data/oracles.md:1338]: I don't understand how this works, precisely how we are simulating the current price of SOL to be able to test this. Can you explain it to me?

Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md:1338]: @mikemaccana Here is the explainer - [Anchor Toml reference ](https://www.anchor-lang.com/docs/manifest#test-validator)- 
The test environment clones the Switchboard `SOL/USD` price feed from the devnet to the local test validator.
This cloned feed maintains its last known state, providing a fixed, realistic SOL price for the tests.
Then tests interact with this local copy of the feed, fetching the SOL price as if it were live data.
While the price doesn't update in real time, it allows consistent testing against a known SOL price point. In essence, we are working with a "snapshot" of the real SOL price feed, frozen at the time of cloning when we start testing with `anchor test` command.  For this to work we have added an additional config in `Anchor.toml` file -

```toml
[test]
startup_wait = 5000
shutdown_wait = 2000
upgradeable = false

[test.validator]
bind_address = "0.0.0.0"
url = "https://api.devnet.solana.com"
ledger = ".anchor/test-ledger"
rpc_port = 8899

[[test.validator.clone]] # switchboard-solana devnet programID
address = "SW1TCH7qEPTdLsDHRgPuMQjbQxKdH2aBStViMFnt64f"

[[test.validator.clone]] # switchboard-solana devnet IDL
address = "Fi8vncGpNKbq62gPo56G4toCehWNy77GgqGkTaAF5Lkk"

# This clones the SOL/USD price feed from devnet to local test validator.
[[test.validator.clone]] # switchboard-solana SOL/USD Feed
address = "GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR"
```

This PublicKey used in tests corresponds to the cloned feed in a local environment.

```typescript
const SOL_USD_SWITCHBOARD_FEED = new PublicKey(
  "GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR"
);
```

Let me know if you still have doubts.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 1 above)
Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md]: Yes resolved as per suggestion.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 2 above)
Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md]: Yes resolved as per suggestion.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 3 above)
Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md]: Yes formatted with deleteme file.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 4 above)
Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md]: Yes updated content to make it more clear.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 5 above)
Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md:449]: Yes fixed this by importing - `import { Big } from "@switchboard-xyz/common";`  which is recommended for Oracle data. These are two different `Big` is from the switchboard common library which uses `big.js` and the other is from `bn.js`.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 6 above)
Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md]: Yes formatted with deleteme file.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 8 above)
Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md]: Yes removed this.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 1 above)
Comment (mikemaccana) [reply content/courses/connecting-to-offchain-data/oracles.md]: Sorry @0xCipherCoder I should be more specific.

Don't do:
```
some text

 - Pyth
 - Switchboard
 - Chainlink
 - Dia 

some text

 - Pyth
 - Switchboard
```
Instead, replace all of that with:

```
some text

 - Pyth
 - Switchboard
 - Chainlink
 - Dia 
```

I hope that makes sense.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 5 above)
Comment (mikemaccana) [reply content/courses/connecting-to-offchain-data/oracles.md:450]: OK they're different got it.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 9 above)
Comment (mikemaccana) [reply content/courses/connecting-to-offchain-data/oracles.md:1349]: I understand account cloning, but I don't understand why
`it withdraws from escrow` and `it fails to withdraw while price is below UnlockPrice` both pass when  `SOL_USD_SWITCHBOARD_FEED` is not changing.

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Getting very close, see two comments below.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 9 above)
Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md:1349]: Okay got the concerns. Although  `SOL_USD_SWITCHBOARD_FEED` is not changing we are changing the price in test as given below to simulate the escrow behaviour.

`const unlockPrice = solPrice.plus(PRICE_OFFSET).toNumber();`

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 1 above)
Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md]: Just curious should we then remove description that has been given again for Pyth and Switchboard? Just keep the original oneliner description then go ahead with switchboard.

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 9 above)
Comment (mikemaccana) [reply content/courses/connecting-to-offchain-data/oracles.md:1349]: Aha! Can you please add a comment in the tests to say this?

File: content/courses/connecting-to-offchain-data/oracles.md (hunk 1 above)
Comment (mikemaccana) [reply content/courses/connecting-to-offchain-data/oracles.md]: Yes - keep all the information but use a single list.

Comment (0xCipherCoder) [reply content/courses/connecting-to-offchain-data/oracles.md]: Updated content list as per suggestion.

**Review by mikemaccana — APPROVED** [review]

**Comment by mikemaccana** [comment]
Thanks @0xCipherCoder ! I’m on my phone but will mark this as the winner soon.

**Pull request merged**